lib/Sema/CSBindings.cpp

```diff
@@ -1082,6 +1082,17 @@ bool TypeVariableBinding::attempt(ConstraintSystem &cs) const {
     type = cs.openUnboundGenericType(type, dstLocator);
     type = type->reconstituteSugar(/*recursive=*/false);
   }
+  
+  // The are some cases where attempting to infer function types
+  // for a generic parameter have more than one valid viable
+  // bindings e.g. {(supertypes of) () -> Void; (subtypes of)
+  // @convention(block) () -> Void}. In order to disambiguate
+  // this case we increase the score for subtype bindings.
+  if (Binding.Kind == ConstraintSystem::AllowedBindingKind::Subtypes &&
+      TypeVar->getImpl().hasGenericParameter() &&
+      type->is<FunctionType>()) {
```

Comment:
I don't think that this is quite right - at this point we don't know whether it would require conversion or not and doing it always could sway score in other unrelated cases...

Reply:
We need to figure out how to narrow down the situation here were it's known that generic parameter could either be one function type or another and set score then...

Reply:
That seems like a tough one(at least to me)... because at this point the only thing the only information we have its the current binding being attempted, so at least for where I can see we would have to track the previous successful bindings attempted in the constraint system to be able to know this at this point, or maybe check if the system already has a fixed type assigned to this typevar(not sure if this makes sense in this context, that is assuming a previous subtype binding was successful and the fixed type was set to the tyVar in the system) ... do you think this makes sense in some way? @xedin

Reply:
Also, thank you for taking a look at this :)

Reply:
I was thinking that it might be possible to detect this in `PotentialBindings` when all of the bindings are gathered. In our example we'd have `() -> Void <: $T2 <: @convention(block) () -> Void` so I guess we could simplify it down to just `() -> Void` + increased score for  `ScoreKind::SK_FunctionConversion` if the only difference between function types is their representation...

Reply:
Humm interesting ... let me try that, but one thing I'm not sure is if we could do that only based the representations being different... I mean not all function representations are implicit convertible to each other right? So maybe if this is the case we should check for that...

Reply:
If they aren't convertible then this would effectively do nothing...

Reply:
Right, I still can make it work using this approach and TBH I'm a bit confused about it right now...
When you say, in `PotentialBindings` do you have someplace specific in mind? I was trying to detect this in `ConstraintSystem::PotentialBindings::addPotentialBinding` but increase the score inside the `PotentialBindings` seems not possible ... 
Anyways I think I'll defer it to later this week when I have more time to think about this, probably work on other things meanwhile 👍 

And thank you so much for all the help here @xedin :))

Reply:
> When you say, in PotentialBindings do you have someplace specific in mind?

It has to happen when all of the bindings are gathered for a type variable, so somewhere in `determineBestBindings`. 

> increase the score inside the PotentialBindings seems not possible ...
Increasing score is impossible but we could add a flag to `PotentialBinding` that it would need to get a score adjusted when attempted via `TypeVariableBinding::attempt`.

Reply:
Humm, awesome! Thank you for the insight @xedin :)) 
I'll get back to work on that ASAP and ping you when I have something 👍

Reply:
Sounds good, no rush!

Reply:
@xedin Just push the changes, let me know what you think :))

Reply:
@xedin Just pinging, I remembered we had this opened :)